[greatingdor 19기 함경재] Add: 상품 출력 기능 구현

gyungjae-ham opened the pull request:
수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
• products/views.py 수정

수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
- 중복되는 코드 부분 수정
- code align 부분 수정
- 필요없는 import 값 제거
기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.)
<img width="276" alt="products:categories reponse" src="https://user-images.githubusercontent.com/75306311/115236874-da7fcc80-a156-11eb-9f93-d8d70d06da33.png">
<img width="934" alt="products views response" src="https://user-images.githubusercontent.com/75306311/115237019-069b4d80-a157-11eb-949e-62915aedc500.png">

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

gyungjae-ham committed:
Add: 상품 출력 기능 구현
- 메인페이지 ALL(전체상품) 출력 기능
- category_id filter 기능
- 메인페이지 카테고리 호버 창 메뉴 이름, 카테고리 이름 출력 기능

redsaumon commented on greatingdor/urls.py:
```diff
@@ -13,7 +13,10 @@
 """
-from django.urls import path
+from django.urls import path, include
+from products.views import CategoryView
```
Django 모듈과 내가 만든 클래스를 구분해주세요.

gyungjae-ham replied:
네 알겠습니다!!

redsaumon committed:
Update urls.py

Django 모듈과 내가 만든 클래스를 구분해주세요.

redsaumon commented on greatingdor/urls.py:
```diff
@@ -13,7 +13,11 @@
 
 urlpatterns = [
+    path('products', include('products.urls')),
+    path('categories', CategoryView.as_view()),
```
categories 앱에 view가 하나뿐인가요?

gyungjae-ham replied:
categories가 앱이 아니라 view를 끌어와서 바로 보냈습니다

gyungjae-ham replied:
앱은 products 입니다

redsaumon replied:
한 앱에 여러 view가 있을 경우 통일성을 위해 최상위 url과 해당 앱의 url로 view 경로를 지정하는 것을 추천드립니다.

gyungjae-ham replied:
넵 알겠습니다 수정하겠습니다

redsaumon commented on products/views.py:
```diff
@@ -1,3 +1,55 @@
+        limit = int(page_size * page)
+        offset = int(limit - page_size)
+        
+        result = []
+        for info in infos:
+            images  = info.productimage_set.all()
+            image_list = []
+            for image in images:
+                image_list.append(image.image_url)
+            result.append(
+            {
+            'image': image_list,
+            'price': info.price,
+            'title': info.title,
+            'description': info.description,
+            'category': info.category.name,
+            'isnew': True
+            }
+            )
```
list comprehension을 공부해서 적용시켜 보셔도 좋을 거 같습니다. list를 선언하는 새로운 방법을 익히실 수 있고 일반 for loop보다 속도가 2배 가량 빠릅니다.

gyungjae-ham replied:
넵넵 공부해서 수정하겠습니다

gyungjae-ham committed:
Refactor: urls, views 코드 수정
- greatingdor/urls.py django기능과 사용자 입력 부분 분리
- products/views.py에서 image_url 부분 list comprehension 적용

redsaumon commented on products/views.py:
```diff
@@ -1,3 +1,53 @@
+        result = []
+        for info in infos:
+            images  = info.productimage_set.all()
+            image_list = [image.image_url for image in images]
```
좋습니다~ 바깥의 for문에도 적용시켜보시겠어요?

gyungjae-ham replied:
음.... 안의 list comprehension은 복잡하지 않아서 적용했을 경우 메리트가 많이 느껴졌는데 밖의 list는 
안에 들어갈 내용이 많아서 제가 잘 모르는 건지 너무 복잡하게 느껴지네요 팁이 있을까요??

redsaumon replied:
기존 이중 for문의 형태를 보면 큰 for문이 안의 for문을 포함하고 있었죠. list comprehension도 비슷합니다~ 한 번 고민만 해봐주시고 다른 작업 하셔도 좋아요! 오후에 내려가서 같이 얘기 나눠봐요~

gyungjae-ham replied:
아뇨 연우님 힌트를 얻었어요 제가 공부해서 해결해 놓겠습니다 오셔서 확인만 해주세요

redsaumon commented:
경재님 confilct도 해결해주세요~

gyungjae-ham committed:
Merge branch 'master' into feature/products

gyungjae-ham committed:
Refactor: products/views.py 코드수정
- result list comprehension
- conflict 해결

gyungjae-ham committed:
Merge branch 'master' of https://github.com/wecode-bootcamp-korea/19-1st-greatingdor-backend

gyungjae-ham committed:
Merge branch 'master' into feature/products

gyungjae-ham committed:
Merge branch 'feature/products' of https://github.com/wecode-bootcamp-korea/19-1st-greatingdor-backend into feature/products

redsaumon commented on greatingdor/urls.py:
```diff
@@ -16,5 +16,8 @@
 
 urlpatterns = [
-        path('members', include('members.urls')),
+    path('products', include('products.urls')),
+
+
+    path('members', include('members.urls')),
```
이 간격은 뭐죠?

gyungjae-ham replied:
아 merge 할 때 저 부분을 합치면 에러가 발생하더라구요

redsaumon commented on products/views.py:
```diff
@@ -1,3 +1,54 @@
+class CategoryView(View):
+    def get(self, request):
+        menus = Menu.objects.all()
+        result = []
+        
+        for menu in menus:
+            categories = menu.category_set.all()
+            category_list = []
+            for category in categories:
+                category_list.append(category.name)
+            result.append(
+            {
+            'menu': menu.name,
+            'category': category_list,
+            }
+            )
```
경재님 list comprehension 적용 아주 좋습니다~ 이 쪽도 가시죠

gyungjae-ham replied:
오호.....이런걸 또 찾아 내셨군요....오호라

gyungjae-ham committed:
Refactor: products/views.py 코드수정
- list comprehension
- urls.py 빈칸 삭제

redsaumon commented on products/views.py:
```diff
@@ -1,3 +1,46 @@
+
+class CategoryView(View):
+    def get(self, request):
+        menus = Menu.objects.all()
```
경재님 깔끔하고 좋습니다~ 그런데 이 부분도 굳이 변수화시킬 필요 없지 않을까요?

gyungjae-ham replied:
넵 확인해보겠습니다

redsaumon commented on products/views.py:
```diff
@@ -1,3 +1,46 @@
+    def get(self, request):
+        category_id = request.GET.get('category', None)
+
+        if category_id != None:
+            products = Product.objects.filter(category_id=category_id)
+            infos = products.order_by('description')
+        else:    
+            products = Product.objects.all()
+            infos = products.order_by('description')
```
경재님 infos를 만들어주는 부분은 완전히 일치하는데 따로 빼주는 건 어떨까요?

gyungjae-ham replied:
넵 알겠습니다

redsaumon reviewed (commented):
경재님 아주 좋습니다~ 회사나 팀마다 따르는 컨벤션이 있는데 위코드도 컨벤션이 있습니다. 노션 wecode Python Coding Convention의 Code Align 부분 참고해주세요:)

gyungjae-ham committed:
Refactor: products/views.py 코드수정
- 중복되는 코드 부분 수정
- code align 부분 수정

gyungjae-ham committed:
Refactor: 필요없는 import 값 제거

redsaumon commented:
:+1:
